proto/substrait/algebra.proto

```diff
@@ -228,6 +228,20 @@ message AggregateRel {
   }
 }
 
+// The Window relational operate on a group of rows, referred to as a window, and calculate a return value for each row based on the group of rows.
+message WindowRel {
+  RelCommon common = 1;
+  Rel input = 2;
+  repeated Measure measures = 3;
+  repeated Expression partition_expressions = 4;
+  repeated SortField sorts = 5;
```

Comment:
Should a window relation specify sorts?  Or should it just use whatever orderedness is currently established?  For example, the `FetchRel` does not have "sorts".

Reply:
Window relation need to control the way that rows in a partition are ordered, determining the position of the given row in its partition. Refers to the definition of Window in [Spark ](https://spark.apache.org/docs/latest/sql-ref-syntax-qry-select-window.html)and [VELOX](https://facebookincubator.github.io/velox/develop/operators.html#windownode)

Reply:
Yes, I agree that a window relation needs an established ordering.  I don't know that the window relation itself has to establish the ordering.

A `FetchRel` also relies on an established ordering.  However, it does not establish the ordering itself.

For example, if you wanted to do a fetch to limit the rows to 10Mi and then run a window operation on that you should be able to do:

```
OrderBy -> Fetch -> Window (no sorting)
```

Curious for @jacques-n or @cpcloud 's input on this as they have more experience with multiple implementations.

However, I will concede that the two established systems you mentioned (spark and velox) would not support that and so maybe we should do something more traditional here.

Reply:
Also, how is this related to the `sorts` property on `WindowFunction`?

https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L819

Reply:
It may be more reasonable to put the `sorts `property in WindowRel, because the sort requires all window functions to be sorted according to the specified sort order when processing. And we may can remove this sorts property in WindowFunction.

Reply:
SQL allows for each window function to have their own sort:

```
SELECT RANK() OVER (ORDER BY val ASC), RANK() OVER (ORDER BY val DESC) FROM my_table;
```

Would you envision this as two different `WindowRel`s?

Reply:
What is the motivation for a separate window relation? When would I use a `WindowRel` versus a projection with `WindownFunction` expressions?

Reply:
Physically I think there is some advantage in having a a window rel which has one or more window functions that all share the same sorting & partitioning.  I'm not sure how best to mark this as a physical node instead of a logical node (it's also not how I read this originally).  E.g. this matches Velox's interpretation: https://github.com/facebookincubator/velox/blob/main/velox/core/PlanNode.h#L1687-L1689

> /// PlanNode used for evaluating Sql window functions.
> /// All window functions evaluated in the operator have the same
> /// window spec (partition keys + order columns).

It also seems to match Spark's interpretation.

I agree, however, that there is no need for this in a purely logical plan.

Reply:
Then I think I'm not understanding how the logical plan of

```
SELECT
  SUM(x) OVER (PARTITION BY a),
  AVG(y) OVER (PARTITION BY b)
FROM t
```

turns into a physical plan containing `WindowRel`s.

Reply:
Since the partition is different there would be two `WindowRel` nodes.  They could run in parallel or sequentially.  Ideally an engine could find some way to do something to avoid double-accumulating all the data.  Perhaps an order by `a,b` for example.

However, naively, it seems spark SQL gives us:

```
AdaptiveSparkPlan isFinalPlan=false
+- Project [sum(x) OVER (PARTITION BY a ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#63L, sum(y) OVER (PARTITION BY b ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#64L]
   +- Window [sum(y#43) windowspecdefinition(b#45, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS sum(y) OVER (PARTITION BY b ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#64L], [b#45]
      +- Sort [b#45 ASC NULLS FIRST], false, 0
         +- Exchange hashpartitioning(b#45, 200), ENSURE_REQUIREMENTS, [plan_id=42]
            +- Project [y#43, b#45, sum(x) OVER (PARTITION BY a ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#63L]
               +- Window [sum(x#42) windowspecdefinition(a#44, specifiedwindowframe(RowFrame, unboundedpreceding$(), unboundedfollowing$())) AS sum(x) OVER (PARTITION BY a ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)#63L], [a#44]
                  +- Sort [a#44 ASC NULLS FIRST], false, 0
                     +- Exchange hashpartitioning(a#44, 200), ENSURE_REQUIREMENTS, [plan_id=37]
                        +- Project [_1#33 AS x#42, _3#35 AS a#44, _2#34 AS y#43, _4#36 AS b#45]
                           +- SerializeFromObject [knownnotnull(assertnotnull(input[0, scala.Tuple4, true]))._1 AS _1#33, knownnotnull(assertnotnull(input[0, scala.Tuple4, true]))._2 AS _2#34, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, scala.Tuple4, true]))._3, true, false, true) AS _3#35, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, knownnotnull(assertnotnull(input[0, scala.Tuple4, true]))._4, true, false, true) AS _4#36]
                              +- Scan[obj#32]
```

Stripping away the noise:

```
AdaptiveSparkPlan isFinalPlan=false
+- Project [...]
   +- Window [...],
      +- Sort [...],
         +- Exchange,
            +- Project [...],
               +- Window [...],
                  +- Sort [...],
                     +- Exchange,
                        +- Project [...],
                           +- SerializeFromObject [...]
                              +- Scan[..]
```

Reply:
SQL server does something considerably more complicated:

![image](https://user-images.githubusercontent.com/1696093/223580077-6d2e4572-3e80-4c97-8478-614b2be54c43.png)

Reply:
I see, so it's a big chain of projections in the spark case. Thanks, this makes sense to me now.

Reply:
And for completeness, DuckDb:

```
┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│   sum(read_csv.x) OVER    │
│ (PARTITION BY read_csv.a) │
│   avg(read_csv.y) OVER    │
│ (PARTITION BY read_csv.b) │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             #0            │
│             #1            │
│             #2            │
│             #3            │
│             #4            │
│             #5            │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│           WINDOW          │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│avg(y) OVER (PARTITION BY b│
│             )             │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│           WINDOW          │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│sum(x) OVER (PARTITION BY a│
│             )             │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│             #0            │
│             #2            │
│             #1            │
│             #3            │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│          READ_CSV         │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│           EC: 42          │
└───────────────────────────┘                             

```

Reply:
> SQL allows for each window function to have their own sort:
> 
> ```
> SELECT RANK() OVER (ORDER BY val ASC), RANK() OVER (ORDER BY val DESC) FROM my_table;
> ```
> 
> Would you envision this as two different `WindowRel`s?

Yes. Spark will create two different Window operator with different sort spec.

![image](https://user-images.githubusercontent.com/11972570/223607778-9db6afea-e62d-48d8-a938-0d8de64f4a00.png)

Reply:
@westonpace @cpcloud Do you have any further comment about Window Rel?